### PR TITLE
Survey indexing

### PR DESCRIPTION
**Pull request by alexandrkuvaev**
* Implement two indexer classes to speed up access to rows of `Survey.headers` by their indices. Improves performance of `get_gather` and `sample_gather` methods.
* Now `Survey` class implements `indices` property to match `SeismicIndex` interface and `n_gathers` and `indexed_by` properties. `survey.info()` now won't hang for large surveys.
* `Gather` now implements `indexed_by` and `index` properties.
* Remove a dirty performance hack in plot geometry which used a `SeismicIndex` instead of a `Survey` to get a gather on click.
* Switch to inplace `DataFrame` processing where possible and avoid unnecessary copying of survey data.
* Refactor `Survey.construct_attribute_map` method.
* Add an `evaluate` method to metric map similar to that of `MetricsAccumulator`, simplify classes of metric maps.

### CODE REVIEW

File: seismicpro/src/survey/survey.py [hunk 1]
```diff
@@ -485,7 +517,7 @@ def load_gather(self, headers, limits=None, copy_headers=True):
         """
         if copy_headers:
             headers = headers.copy()
-        trace_indices = headers.reset_index()["TRACE_SEQUENCE_FILE"].values - 1
+        trace_indices = get_cols(headers, "TRACE_SEQUENCE_FILE").ravel() - 1
```
[the remark this answers is not in the document]
Comment (alexandrkuvaev) [reply seismicpro/src/survey/survey.py]: Renamed

File: seismicpro/src/survey/survey.py [hunk 2]
```diff
@@ -628,27 +654,37 @@ def load_first_breaks(self, path, trace_id_cols=('FieldRecord', 'TraceNumber'),
         first_breaks_df = pd.read_csv(path, delimiter=delimiter, names=file_columns,
                                       decimal=decimal, encoding=encoding, **kwargs)
 
-        headers = self.headers.reset_index()
-        headers = headers.merge(first_breaks_df, on=trace_id_cols)
+        headers = self.headers.merge(first_breaks_df, on=trace_id_cols)
```
[the remark this answers is not in the document]
Comment (alexandrkuvaev) [reply seismicpro/src/survey/survey.py]: Fixed

File: seismicpro/src/survey/survey.py [hunk 3]
```diff
@@ -909,26 +947,20 @@ def generate_supergathers(self, size=(3, 3), step=(20, 20), modulo=(0, 0), reind
...
+        supergather_lines = pd.DataFrame(create_supergather_index(supergather_centers, size),
                                          columns=super_line_cols+line_cols)
-        self.headers = pd.merge(supergather_lines, headers, on=line_cols)
 
-        if reindex:
-            index_cols = super_line_cols
-        self.headers.set_index(index_cols, inplace=True)
-        self.headers.sort_index(kind="stable", inplace=True)
+        headers = pd.merge(supergather_lines, self.headers, on=line_cols)
+        headers.set_index(index_cols, inplace=True)
```
[the remark this answers is not in the document]
Comment (alexandrkuvaev) [reply seismicpro/src/survey/survey.py:965]: Fixed

**Review by dpodvyaznikov — APPROVED** [review]

File: seismicpro/src/utils/indexer.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    """
+    def __init__(self, index):
+        super().__init__(index)
+        unique_indices_pos = np.where(~index.duplicated())[0]
+        ix_start = unique_indices_pos
+        ix_end = chain(unique_indices_pos[1:], [len(index)])
+        self.unique_indices = index[unique_indices_pos]
+        self.index_to_headers_pos = {ix: range(*args) for ix, *args in zip(self.unique_indices, ix_start, ix_end)}
+
+    def get_loc(self, index):
```
Comment (nikita-klsh) [inline seismicpro/src/utils/indexer.py:62]:
```suggestion
    def get_loc(self, indices):
```
Please consider to rename to unify with `get_headers_by_indices`, not mess with `self.index` and indicate that it's iterable

File: seismicpro/src/utils/indexer.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+class TraceIndexer(BaseIndexer):
+    """Construct an indexer for unique monotonic `index`. Should not be instantiated directly, use `create_indexer`
+    function instead.
+    """
+    def __init__(self, index):
+        super().__init__(index)
+        self.unique_indices = index
+        _ = self.get_loc(index[:1])  # Warmup of `get_loc`: the first call is way slower than the following ones
+
+    def get_loc(self, index):
```
Comment (nikita-klsh) [inline seismicpro/src/utils/indexer.py:34]:
```suggestion
    def get_loc(self, indices):
```

**Review by nikita-klsh — APPROVED** [review]

File: seismicpro/src/utils/indexer.py (hunk 4 above)
Comment (alexandrkuvaev) [reply seismicpro/src/utils/indexer.py:62]: Renamed

File: seismicpro/src/utils/indexer.py (hunk 5 above)
Comment (alexandrkuvaev) [reply seismicpro/src/utils/indexer.py:34]: Renamed

File: seismicpro/src/survey/survey.py [hunk 6]
```diff
@@ -968,18 +1013,36 @@ def plot_geometry(self, **kwargs):
         """
         SurveyGeometryPlot(self, **kwargs).plot()
 
-    def construct_attribute_map(self, attribute, by, agg=None, bin_size=None, **kwargs):
+    def construct_attribute_map(self, attribute, by, drop_duplicates=False, agg=None, bin_size=None, **kwargs):
         """Construct a map of trace headers values aggregated by gathers.
```
[the remark this answers is not in the document]
Comment (alexandrkuvaev) [reply seismicpro/src/survey/survey.py:1017]: Fixed

File: seismicpro/src/survey/survey.py [hunk 7]
```diff
@@ -30,7 +30,7 @@ class Survey:  # pylint: disable=too-many-instance-attributes
     method.
 
     The resulting gather type depends on `header_index` argument, passed during `Survey` creation: traces are grouped
-    into gathers by the common value of headers, defined by `header_index`. Some usual values of `header_index`
+    into gathers by the common value of headers, defined by `header_index`. Some common values of `header_index`
```
[the remark this answers is not in the document]
Comment (alexandrkuvaev) [reply seismicpro/src/survey/survey.py]: Changed to frequently used

File: seismicpro/src/metrics/metric_map.py [hunk 8]
```diff
@@ -64,11 +64,47 @@ def y_tick_labels(self):
...
+        -------
+        metric_val : float
+            Evaluated metric value.
+        """
+        if agg is None:
+            agg = self.agg
+        return self.metric_data[self.metric_name].agg(agg)
+
+    def get_worst_coords(self, is_lower_better=None):
+        """Get coordinates with the worst metric value dependind on `is_lower_better`. If not given, `is_lower_better`
```
[the remark this answers is not in the document]
Comment (alexandrkuvaev) [reply seismicpro/src/metrics/metric_map.py]: Fixed

**Pull request merged**